### PR TITLE
Catch and display global errors in mitt-konto

### PR DESCRIPTION
**Pull request by zudov**
2018-11-13-101155_884x419_scrot

- [X] timeout requests appropriately
- [X] have a view area to show the error messages
- [X] capture network errors
- [ ] capture 5XX responses
- [ ] translate messages to Swedish

### CODE REVIEW

File: apps/mitt-konto/src/MittKonto/Main.purs [hunk 1]
```diff
@@ -306,7 +320,16 @@ loginView { state, setState } = React.fragment
...
                 setState $ setLoggedInUser $ Just user
-          , launchAff_: Aff.launchAff_ <<< withSpinner (setState <<< setLoading)
+          , launchAff_: Aff.runAff_
+              (case _ of
+                 Left err
+                   | Just errMessage <- Persona.networkError err -> do
+                       setState $ setErrorMessage $ Just
+                         "Failed to connect. Check your connection and try again later."
+                   | otherwise -> do
+                       setState $ setErrorMessage $ Just $ Error.message err
```
[the remark this answers is not in the document]
Comment (f-f) [reply apps/mitt-konto/src/MittKonto/Main.purs]: Agreed, let's display something generic, and log the actual error (we actually want to collect them in analytics, I'll open an issue for that)

File: apps/mitt-konto/resources/mitt-konto.less [hunk 2]
```diff
@@ -201,6 +201,15 @@ h2.mitt-konto---heading {
...
 
+.mitt-konto--error-message {
+    padding: 0.8em;
+    font-weight: 300;
+    font-size: 1.3em;
+    border-radius: 0 0 0.5em 0.5em;
+    color: @grey-deepdark;
+    background-color: salmon;
+}
+
```
Comment (f-f) [inline apps/mitt-konto/resources/mitt-konto.less]: We should use stylings from [our styleguide](https://hbl.stage.ksf.infomaker.io/styleguide-2/), e.g. in this case we should use an error notification box:
image

**Comment by vapaj** [comment]
Cool stuff 👍

File: apps/mitt-konto/src/MittKonto/Main.purs (hunk 1 above)
[the remark this answers is not in the document]
Comment (f-f) [reply apps/mitt-konto/src/MittKonto/Main.purs]: Issue for that is open: #30

**Comment by zudov** [comment]
This can now be reviewed.

I've moved the error view in a separate package `ksf-alert` that renders alerts. Currently it's just a stateless view for a single message, but in the future it can handle:

- showing several alerts
- dismiss them by clicking
- dismiss automatically after some time
- provide some helpers for adding it to the top component and sending alerts from child components

**Comment by zudov** [comment]
Translations are still needed. The 5xx errors would go in another merge request (@vapaj is working on it), they turn out to be a bit more context-dependent and also need some children<->parent communication.

**Comment by zudov** [comment]
New look:

2018-11-16-181621_783x449_scrot

File: apps/mitt-konto/src/MittKonto/Main.purs [hunk 3]
```diff
@@ -76,16 +75,17 @@ app = React.component
     render { state, setState } =
       React.fragment
         [ navbarView { state, setState }
-        , alertView state.alert
-        , classy DOM.div "clearfix"
+        , classy DOM.div "mt4 mb4"
+            [ foldMap alertView state.alert ]
+        , classy DOM.div "mt4 mb4 clearfix"
```
Comment (zudov) [inline apps/mitt-konto/src/MittKonto/Main.purs]: @f-f the margins are now here. They get nicely collapsed and things are even.

File: apps/mitt-konto/src/MittKonto/Main.purs [hunk 4]
```diff
@@ -321,7 +321,7 @@ loginView { state, setState } = React.fragment
                 log "Fetching user succeeded"
                 setState $ setLoggedInUser $ Just user
           , launchAff_:
-              Aff.runAff_ (either (setState <<< setAlert <<< errorAlert) pure)
+              Aff.runAff_ (setState <<< setAlert <<< either errorAlert (const Nothing))
```
Comment (zudov) [inline apps/mitt-konto/src/MittKonto/Main.purs]: Here I've added @vapaj's suggestion to remove the error message once the aff succeeds at anything, this isn't the perfect handling, but it handles most of the use cases and is very straightforward.

File: apps/mitt-konto/src/MittKonto/Main.purs [hunk 5]
```diff
@@ -331,6 +347,21 @@ loginView { state, setState } = React.fragment
             ]
         ]
 
+errorAlert :: Error -> Maybe Alert
+errorAlert err = oneOf
+  [ do { method, url } <- Persona.networkError err
+       pure
+         { level: Alert.danger
+         , title: "Failed to connect."
+         , message: "Check your connection and try again later."
```
Comment (f-f) [inline apps/mitt-konto/src/MittKonto/Main.purs]: I think we were waiting for the translated Swedish texts, should we have them in this PR or in a new one?

**Pull request merged**